### ISSUE TITLE
`persistent` clear foreign tables on automatic migration due to cascading delete

### ISSUE THREAD

**Issue by KtorZ** [cardano-foundation/cardano-wallet#1279]
# Context
 
 

| Information  | -                                                     |
| ---          | ---                                                   |
| Version      | All  |
| Platform     | All           |
| Installation | N/A             |

See https://input-output-rnd.slack.com/archives/CGH3J2Q1L/p1579253925019600

# Steps to Reproduce

1. Build master.
2. Start the wallet backend with cardano-wallet-jormungandr launch ... 
3. Create a brand new wallet.
4. Stop the wallet backend.
5. Use sqlite3 to manually observe the contents of the seq_state_address table (with something like select distinct(slot) from seq_state_address;).
6. You should see that it typically has ~11 entries (genesis + 10 checkpoints) entries.
7. Add a new column foo to the checkpoint table definition in lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs, giving it a type of Maybe Word32. (Persistent should be able to make an automatic migration for this case, as we've used Maybe.)
8. Rebuild, and then restart the wallet backend.
9. Observe in the logs that persistent has performed several automatic migrations.
10. Stop the wallet backend.

## Expected behavior

11. Now use sqlite3 to look at the entries of the seq_state_address table. Previous entries should still be there (modulo the checkpoints moving forward) and in particular the entry for the genesis checkpoint should remain!

## Actual behavior
 

11. Now use sqlite3 to look at the entries of the seq_state_address table. All previous entries are gone.

---

# Resolution 

- One possible approach: forcing a manual rollback to genesis whenever there are migrations applied to the database, regardless of the migration. Only the foreign tables are affected, so static metadata and wallet credentials are preserved. This mean that we would force resync of wallets on every updates that have database changes, but that's a fair price to pay for now. ==> #1290 

---

# QA 

See comment of https://github.com/input-output-hk/cardano-wallet/pull/1290#issue-364947415

**Comment by jonathanknowles** [comment]
@KtorZ 

Just as an aside, the persistent manual recommends against the use of automatic migrations in production environments:

> Using automated database migrations is only recommended in development environments. Allowing your application to modify your database schema in a production environment is _very strongly discouraged_.

(source: https://www.yesodweb.com/book/persistent)

**Comment by KtorZ** [comment]
Perhaps that is something they could highlight even more, especially in the haddock documentation :sweat_smile:  ... Yet, it's open-source, so we might as well give them a little PR.

**Comment by piotr-iohk** [comment]
lgtm

**Comment by piotr-iohk** [comment]
I've encountered 0 balance issue, can reproduce with local integration node:
 - start 2020.01.14
 - restore faucet byron wallet (it has some funds) -> `arctic decade pink easy jar index base bright vast ocean hard pizza`
 - stop wallet and start 2020.01.21 on the same data
 - restored byron wallet balance = 0

The same behavior observed for shelley faucet wallet -> `vintage poem topic machine hazard cement dune glimpse fix brief account badge mass silly business
`

**Comment by KtorZ** [comment]
I spent some time looking into this and I believe that in order to fix this properly, we need to:

a) revert the change from #1290 and 
b) remove persistent automated migration and have the migration done entirely "manually". 

We can inspect the Sqlite database using the SQLite metatables to see whether tables or columns exists. This would give us much more control on what's actually happening and would be a better experience than this awkward rollback to genesis in case of migrations  (cc @jonathanknowles)

**Comment by piotr-iohk** [comment]
In the meantime the workaround in case of encountering the issue after upgrading to new `cardano-wallet` is to delete and restore the wallet again using mnemonic sentence.

**Comment by piotr-iohk** [comment]
Looking at it after #1340.

Testing `2019-12-23` -> `master`.
 - in `2019-12-23` create faucet byron wallet with funds
 - run `master` on the same state-dir
OK: My Byron wallet has funds after migration :tada: .

However I don't see any entry:
```
n migrations were applied to the database.
```
I was expecting some migrations between `2019-12-23` and now... :thinking:

**Comment by piotr-iohk** [comment]
More details of the above :point_up_2: .

**TL/DR:**
 - migration `2019-12-23` -> `2020-01-27`
  migrated wallet has balance 0. Logs indicate there was a migration.

- migration `2019-12-23` -> `master`
  migrated wallet has correct balance, the logs do not say there was a migration.

**Step by step:**

- start wallet [v2019-12-23](https://github.com/input-output-hk/cardano-wallet/releases/tag/v2019-12-23) 

```
cd lib/jormungandr/test/data/jormungandr/test_scripts
cardano-wallet-2019-12-23 launch --genesis-block ../block0.bin --node-port 8080 --state-dir ./data_launch -- --secret ../secret.yaml --config ../config.yaml
```

- Create a faucet byron wallet:
```
curl  -vX POST http://localhost:8090/v2/byron-wallets \
  -H "Content-Type: application/json; charset=utf-8" \
  -d '{
"name": "Byron Faucet ",
"mnemonic_sentence": [ "arctic", "decade", "pink", "easy", "jar", "index", "base", "bright", "vast", "ocean", "hard", "pizza"],
"passphrase": "Secure Passphrase"
}' --http1.1  |  jq
```
(wallet created, wallet balance = 1000000000000)


**There are migrations in the log when you migrate to `2020-01-27`.**
- Run version [v2020-01-27](https://github.com/input-output-hk/cardano-wallet/releases/tag/v2020-01-27) on the same --state-dir as [v2019-12-23](https://github.com/input-output-hk/cardano-wallet/releases/tag/v2019-12-23):
```
cardano-wallet-2020-01-27 launch --genesis-block ../block0.bin --node-port 8080 --state-dir ./data_launch -- --secret ../secret.yaml --config ../config.yaml
```

In the log there is:

```
[cardano-wallet.application:Info:22] [2020-02-13 10:32:44.23 UTC] Found existing wallet: aa129a07d1ce083e67597348f1788747a034686e
[cardano-wallet.pools-engine:Info:27] [2020-02-13 10:32:44.23 UTC] Monitoring stake pools. Currently at b3b74ab7-[1269394.9#5]
[cardano-wallet.wallet-db:Notice:28] [2020-02-13 10:32:44.25 UTC] 48 migrations were applied to the database.
[cardano-wallet.wallet-db:Notice:28] [2020-02-13 10:32:44.25 UTC] Some automated database migrations happened. Forced to roll back to genesis.
[cardano-wallet.main:Info:22] [2020-02-13 10:32:44.26 UTC] Wallet backend server listening on 127.0.0.1:8090

```
The wallet has incorrect balance after migration = 0.

**However when you migrate to recent master there is no indication in the log about the migrations.**
 - run `master` on the same --state-dir as [v2019-12-23](https://github.com/input-output-hk/cardano-wallet/releases/tag/v2019-12-23)
```
cardano-wallet-jormungandr launch --genesis-block ../block0.bin --node-port 8080 --state-dir ./data_launch -- --secret ../secret.yaml --config ../config.yaml
```

In the log there is (no indcation about any db migration):
```
[cardano-wallet.application:Info:22] [2020-02-13 10:31:20.11 UTC] Found existing wallet: aa129a07d1ce083e67597348f1788747a034686e
[cardano-wallet.pools-engine:Info:27] [2020-02-13 10:31:20.12 UTC] Monitoring stake pools. Currently at 38975fbb-[1269389.9#9]
[cardano-wallet.main:Info:22] [2020-02-13 10:31:20.13 UTC] Wallet backend server listening on 127.0.0.1:8090
```

Ofc here, the wallet has correct balance after migration = 1000000000000.

**Comment by piotr-iohk** [comment]
Ok, according to @KtorZ:

_So. There's indeed no DB migration since v2019-12-23 ....
v2020-01-27 had the first version of the fix which removed foreign key references and other stuff.. Which.. induced 48 migrations. But, the new fix we merged recently actually revert that, and works differently, so no more migrations are needed._

**Comment by piotr-iohk** [comment]
Ok, tested also on:
`v2019-12-09` -> `master`
 - there are migrations between those versions:
```
[cardano-wallet.pools-db:Error:22] [2020-02-13 15:53:55.84 UTC] Failed to migrate the database: : NOT NULL constraint failed: pool_owner_backup.pool_owner_index
[cardano-wallet.pools-db:Notice:22] [2020-02-13 15:53:55.84 UTC] Non backward compatible database found. Removing old database and re-creating it from scratch. Ignore the previous error.
[cardano-wallet.pools-db:Notice:22] [2020-02-13 15:53:55.88 UTC] 5 migrations were applied to the database.
[cardano-wallet.application:Info:22] [2020-02-13 15:53:55.88 UTC] Found existing wallet: aa129a07d1ce083e67597348f1788747a034686e
[cardano-wallet.wallet-db:Notice:28] [2020-02-13 15:53:55.88 UTC] checkpoint table does not contain required field 'active_slot_coeff'. Adding this field with a default value of 1.0.
[cardano-wallet.pools-engine:Info:27] [2020-02-13 15:53:55.89 UTC] Monitoring stake pools. Currently at genesis
[cardano-wallet.wallet-db:Notice:28] [2020-02-13 15:53:55.91 UTC] 6 migrations were applied to the database.
```
 - also the faucet byron wallet has correct balance after migration
`arctic decade pink easy jar index base bright vast ocean hard pizza`

I have made some sanity testing after migration and wallet seems to work fine :tada: 
 - create/delete wallet
 - transactions
 - delegation
